### ISSUE TITLE
master, 1.19.0-alpha.1 terraform output launch_templates user-data not base64 encoded

### ISSUE THREAD

**Issue by fred-vogt** [kubernetes/kops#9518]
* [release-1-18 branch](https://github.com/kubernetes/kops/tree/release-1.18) **works**:
  * [upup/pkg/fi/cloudup/awstasks/launchtemplate_target_terraform.go ~ line 210](https://github.com/kubernetes/kops/blob/1b7eb3cb4d19f39f0c6d1d0bfb35e82e591cc409/upup/pkg/fi/cloudup/awstasks/launchtemplate_target_terraform.go#L210)
* master / 1-19   **featureflag.Terraform012.Enabled()**
  * [upup/pkg/fi/cloudup/awstasks/launchtemplate_target_terraform.go ~ lineL211](https://github.com/kubernetes/kops/blob/b13209d65456153fd9ff196c850168e60ee46d09/upup/pkg/fi/cloudup/awstasks/launchtemplate_target_terraform.go#L211)
  *  [commit - Only store launchtemplate userdata in plaintext for terraform 0.12+](https://github.com/kubernetes/kops/commit/c5882446c8052b32520e0e86e5f292a9bc56557f#diff-2d44d7e310fb2d36340d9f29d5650286) - seems to be the cause ?
    * verdict - yes [Terraform012 is on by default](https://github.com/kubernetes/kops/blob/b13209d65456153fd9ff196c850168e60ee46d09/pkg/featureflag/featureflag.go#L92)
  * TF AWS provider docs [aws_launch_template](https://www.terraform.io/docs/providers/aws/r/launch_template.html):
    ```
    user_data - The Base64-encoded user data to provide when launching the instance.
    ```

KOPS `... update --target=terraform launch templates` (using TF v0.12.28)
```
"InvalidUserData.Malformed: Invalid BASE64 encoding of user data"
```

TL;DR - TF `aws_launch_template` user-data files - `data/aws_launch_template_master-<zone>.{masters,nodes}.<cluster-name>_user_data` aren't base64 encoded anymore

TF error:
```
Error: InvalidUserData.Malformed: Invalid BASE64 encoding of user data.
	status code: 400, request id: <uuid>

  on kubernetes.tf line NNN, in resource "aws_launch_template" "master-<zone>-masters-<cluster-name>":
 NNN: resource "aws_launch_template" "master-<zone>-masters-<cluster-name>" {
```

**1. What `kops` version are you running? The command `kops version`, will display
 this information.**
v1.19.0-alpha.1

**2. What Kubernetes version are you running? `kubectl version` will print the
 version if a cluster is running or provide the Kubernetes version specified as
 a `kops` flag.**
N/A.

**3. What cloud provider are you using?**
AWS.

**4. What commands did you run?  What is the simplest way to reproduce this issue?**
```
kops update cluster --name=... --state=... --out=. --target=terraform
terraform apply ...
```

**5. What happened after the commands executed?**
```
Error: InvalidUserData.Malformed: Invalid BASE64 encoding of user data.
	status code: 400, request id: ad50838e-bee3-4b5a-a032-a537bd84ba70

  on kubernetes.tf line ..., in resource "aws_launch_template" "master-<zone>-masters-<cluster-name>":
 479: resource "aws_launch_template" "master-<zone>-masters-<cluster-name>" {
```

**6. What did you expect to happen?**
Previous kops versions (1-18b2 ?) didn't have this issue when using `KOPS_FEATURE_FLAGS=EnableLaunchTemplates`.
 
**7. Please provide your cluster manifest. Execute
  `kops get --name my.example.com -o yaml` to display your cluster manifest.
  You may want to remove your cluster name and other sensitive information.**
N/A.

**8. Please run the commands with most verbose logging by adding the `-v 10` flag.
  Paste the logs into this report, or in a gist and provide the gist link here.**
N/A.

**9. Anything else do we need to know?**
With previous working versions of kops (1.18-b1/b2 ?) the files in:
```
./data/aws_launch_template_master-<zone>.{masters,nodes}.<cluster-name>_user_data
```
Were `base64` encoded.
With v1.19.0-alpha.1 the files are plain text.

**Comment by fred-vogt** [comment]
@rifelpet - Thanks for your efforts my friend.  This is the ONLY issue I've found so far with the [kops - v1.19.0-alpha.1](https://github.com/kubernetes/kops/releases/tag/v1.19.0-alpha.1).

This is with [kubernetes - v1.19.0-beta.2](https://github.com/kubernetes/kubernetes/releases/tag/v1.19.0-beta.2).

**Comment by rifelpet** [comment]
Hi @fred-vogt thanks for the report. Can you paste the aws_launch_template resource definition? It should be using `filebase64()` which will base64 encode the contents of the file, so that the file is stored in plaintext and the string passed to the user_data argument is base64 encoded.

**Comment by fred-vogt** [comment]
@rifelpet - Will do.  Spoiler alert `kops update` is not using `filebase64`.
I love the approach of keeping the on disk files "plain" and using `filebase64(...)` in the terraform output.

I'll post the slightly redacted `kubernetes.tf` in a few minutes ...

**Comment by fred-vogt** [comment]
@rifelpet - it must be happening in [upup/pkg/fi/cloudup/awstasks/launchtemplate_target_api.go](https://github.com/kubernetes/kops/blob/b13209d65456153fd9ff196c850168e60ee46d09/upup/pkg/fi/cloudup/awstasks/launchtemplate_target_api.go#L260)

**Comment by johngmyers** [comment]
It appears that `writeLiteral()` in pkg/fi/cloudup/terraform/hcl2.go knows nothing of filebase64. The code in `literal.LiteralFileExpression()` only puts filebase64 in `Value`, which is only used for terraform 0.11.

**Comment by rifelpet** [comment]
Ah you're right, I'll try to get a fix in place soon
/assign
